### PR TITLE
Display non-blocking warnings for synchronization delays

### PR DESCRIPTION
**Pull request by TimothyAsirJeyasing**
https://issues.redhat.com/browse/RHSTOR-5895

### CODE REVIEW

**Comment by TimothyAsirJeyasing** [comment]
image

File: packages/mco/components/modals/app-failover-relocate/failover-relocate-modal.tsx [hunk 1]
```diff
@@ -58,9 +69,26 @@ export const FailoverRelocateModal: React.FC<FailoverRelocateModalProps> = (
...
+  const replicationType = placementControl?.replicationType;
+
+  // Determine if sync is delayed (critical state)
+  const replicationHealth = getReplicationHealth(
+    lastGroupSyncTime,
+    volumesSchedulingInterval,
+    replicationType
+  );
+  const isSyncDelayed =
+    replicationHealth === VOLUME_REPLICATION_HEALTH.CRITICAL;
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/failover-relocate-modal.tsx:87]:
```suggestion
```

Move this logic inside failover-relocate-modal-body.tsx under `validatePlacement` function

Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/failover-relocate-modal.tsx:87]: i think we can check for both `CRITICAL` as well as `WARNING`

File: packages/mco/components/modals/app-failover-relocate/failover-relocate-modal.tsx [hunk 2]
```diff
@@ -169,4 +210,5 @@ export const FailoverRelocateModal: React.FC<FailoverRelocateModalProps> = (
 type FailoverRelocateModalProps = ApplicationProps & {
   isOpen: boolean;
   close?: () => void;
+  drPolicies: DRPolicyKind[];
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/failover-relocate-modal.tsx:213]:
```suggestion
 
```

File: packages/mco/components/modals/app-failover-relocate/parser/argo-application-set-parser.tsx [hunk 3]
```diff
@@ -99,7 +101,10 @@ export const ArogoApplicationSetParser = (
         drLoadError
       )
     );
+
   const aroAppSetResource = aroAppSetResources?.formattedResources?.[0];
+  const drPolicies = drResources?.drPolicies;
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/parser/argo-application-set-parser.tsx:106]:
```suggestion

```

File: packages/mco/components/modals/app-failover-relocate/parser/argo-application-set-parser.tsx [hunk 4]
```diff
@@ -150,6 +156,7 @@ export const ArogoApplicationSetParser = (
             isTargetClusterFenced: isDRClusterFenced(targetDRCluster),
             isPrimaryClusterFenced: isDRClusterFenced(primaryDRCluster),
             areSiblingApplicationsFound: !!siblingApplications?.length,
+            drPolicyName: drPlacementControl.spec.drPolicyRef.name,
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/parser/argo-application-set-parser.tsx:159]:
```suggestion
            schedulingInterval: 
```

File: packages/mco/components/modals/app-failover-relocate/parser/argo-application-set-parser.tsx [hunk 5]
```diff
@@ -166,6 +173,7 @@ export const ArogoApplicationSetParser = (
       loadError={loadError}
       loaded={loaded}
       close={close}
+      drPolicies={drPolicies}
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/parser/argo-application-set-parser.tsx:176]:
```suggestion
     
```

File: packages/mco/components/modals/app-failover-relocate/parser/discovered-application-parser.tsx [hunk 6]
```diff
@@ -94,6 +94,7 @@ export const DiscoveredApplicationParser: React.FC<
     );
 
   const drResource = drResources?.formattedResources?.[0];
+  const drPolicies = drResources?.drPolicies;
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/parser/discovered-application-parser.tsx:97]:
```suggestion
 
```

Comment (GowthamShanmugam) [reply packages/mco/components/modals/app-failover-relocate/parser/discovered-application-parser.tsx:97]: if (loaded && !loadError) {
      const { drClusters, drPlacementControls, drPolicy } = drResource || {};

File: packages/mco/components/modals/app-failover-relocate/parser/discovered-application-parser.tsx [hunk 7]
```diff
@@ -159,6 +160,7 @@ export const DiscoveredApplicationParser: React.FC<
           isDRActionReady: checkDRActionReadiness(drPlacementControl, action),
           snapshotTakenTime: drPlacementControl?.status?.lastGroupSyncTime,
           replicationType: findDRType(drClusters),
+          drPolicyName: drPlacementControl?.spec?.drPolicyRef?.name,
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/parser/discovered-application-parser.tsx:163]:
```suggestion
          schedulingInterval: drPolicy.spec?.schedulingInterval
```

File: packages/mco/components/modals/app-failover-relocate/parser/discovered-application-parser.tsx [hunk 8]
```diff
@@ -178,6 +180,7 @@ export const DiscoveredApplicationParser: React.FC<
       loaded={loaded}
       close={closeModal}
       message={getAlertMessage(action, t)}
+      drPolicies={drPolicies}
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/parser/discovered-application-parser.tsx:183]:
```suggestion
```

File: packages/mco/components/modals/app-failover-relocate/subscriptions/failover-relocate-modal.tsx [hunk 9]
```diff
@@ -102,6 +106,38 @@ export const SubscriptionFailoverRelocateModal: React.FC<FailoverRelocateModalPr
...
+
+        return (
+          getReplicationHealth(
+            drPlacementControl?.status?.lastGroupSyncTime,
+            volumesSchedulingInterval,
+            replicationType
+          ) === VOLUME_REPLICATION_HEALTH.CRITICAL
+        );
+      });
+    };
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/failover-relocate-modal.tsx:137]:
```suggestion

```

Comment (GowthamShanmugam) [reply packages/mco/components/modals/app-failover-relocate/subscriptions/failover-relocate-modal.tsx:137]: a simple way is to set error  state from the peer-cluster-status.tsx, it is already finding worst sync time

File: packages/mco/components/modals/app-failover-relocate/failover-relocate-modal.tsx (hunk 1 above)
Comment (TimothyAsirJeyasing) [reply packages/mco/components/modals/app-failover-relocate/failover-relocate-modal.tsx]: Sure, will add that as well

**Comment by GowthamShanmugam** [comment]
> Screenshot from 2024-10-23 16-45-25 Screenshot from 2024-10-23 16-58-34

Just a doubt, the first screenshot shows an error when no subscription groups are selected, and the second screenshot shows an unknown target cluster. Is this feature tested by changing the code directly?

File: packages/mco/components/modals/app-failover-relocate/failover-relocate-modal-body.tsx [hunk 10]
```diff
@@ -79,30 +81,56 @@ const validatePlacement = (
...
+  if (preCheckError) {
+    return preCheckError;
+  }
+
+  // Check if sibling applications are found
+  if (placementControl?.areSiblingApplicationsFound) {
+    return isFailoverAction
+      ? ErrorMessageType.SIBLING_APPLICATIONS_FOUND_FAILOVER
+      : ErrorMessageType.SIBLING_APPLICATIONS_FOUND_RELOCATE;
+  }
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/failover-relocate-modal-body.tsx:115]: let's move this sibling application check after volume health check, this is less priority one

File: packages/mco/components/modals/app-failover-relocate/helper/error-messages.tsx [hunk 11]
```diff
@@ -269,6 +270,20 @@ export const ErrorMessages = (
...
+    title: t('Inconsistent data on target cluster'),
+    message: (
+      <Trans t={t}>
+        <p>
+          The target cluster&apos;s volumes contain data inconsistencies caused
+          by synchronization delays. Performing the failover could lead to data
+          loss. Refer to the corresponding VolumeSynchronizationDelay OpenShift
+          alert(s) for more information.
+        </p>
+      </Trans>
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/helper/error-messages.tsx:283]: Also, no need <Trans> we can use t('') 

typo here: `cluster&apos;s`

File: packages/mco/components/modals/app-failover-relocate/parser/argo-application-set-parser.tsx [hunk 12]
```diff
@@ -150,6 +160,7 @@ export const ArogoApplicationSetParser = (
             isTargetClusterFenced: isDRClusterFenced(targetDRCluster),
             isPrimaryClusterFenced: isDRClusterFenced(primaryDRCluster),
             areSiblingApplicationsFound: !!siblingApplications?.length,
+            schedulingInterval: drPolicy?.spec.schedulingInterval,
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/parser/argo-application-set-parser.tsx:163]:
```suggestion
            schedulingInterval: drPolicy?.spec?.schedulingInterval,
```

File: packages/mco/components/modals/app-failover-relocate/subscriptions/error-messages.tsx [hunk 13]
```diff
@@ -243,4 +244,18 @@ export const ErrorMessages = (
...
+      <Trans t={t}>
+        <p>
+          The target cluster&apos;s volumes contain data inconsistencies caused
+          by synchronization delays. Performing the failover could lead to data
+          loss. Refer to the corresponding VolumeSynchronizationDelay OpenShift
+          alert(s) for more information.
+        </p>
+      </Trans>
+    ),
+    variant: AlertVariant.warning,
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/error-messages.tsx:259]: same here: https://github.com/red-hat-storage/odf-console/pull/1646/files#r1816354052

File: packages/mco/components/modals/app-failover-relocate/subscriptions/error-messages.tsx [hunk 14]
```diff
@@ -243,4 +244,18 @@ export const ErrorMessages = (
     title: t('No subscription groups are found.'),
     variant: AlertVariant.danger,
   },
+  [ErrorMessageType.SYNC_DELAY]: {
+    title: t('Inconsistent data on target cluster'),
+    message: (
+      <Trans t={t}>
+        <p>
+          The target cluster&apos;s volumes contain data inconsistencies caused
```
Comment (alfonsomthd) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/error-messages.tsx:252]: You meant `cluster's` ?
```suggestion
          The target cluster's volumes contain data inconsistencies caused
```

**Comment by TimothyAsirJeyasing** [comment]
> > Screenshot from 2024-10-23 16-45-25 Screenshot from 2024-10-23 16-58-34
> 
> Just a doubt, the first screenshot shows an error when no subscription groups are selected, and the second screenshot shows an unknown target cluster. Is this feature tested by changing the code directly?

yes, just to see the output message..

File: packages/mco/components/modals/app-failover-relocate/failover-relocate-modal-body.tsx (hunk 10 above)
Comment (TimothyAsirJeyasing) [reply packages/mco/components/modals/app-failover-relocate/failover-relocate-modal-body.tsx:115]: ack

File: packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx [hunk 15]
```diff
@@ -78,25 +83,33 @@ const getPeerStatusSummary = (
...
+          acc.peerReadiness,
+          drpcState,
+          actionType,
+          t
+        ),
+        dataLastSyncedOn: getDataLastSyncTime(acc.dataLastSyncedOn, drpcState),
+        replicationHealth: getReplicationHealth(
+          lastGroupSyncTime,
+          volumesSchedulingInterval,
+          replicationType
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx:107]: it simply takes the last DRPC replication health. It's not correct. it has to take the Non-healthy replication state by comparing with its previous DRPC health.

Comment (GowthamShanmugam) [reply packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx:107]: Pass acc.replicationHealth(previous DRPC sync health) and return the worst one by comparing it with the current DRPC replication health

File: packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx [hunk 16]
```diff
@@ -35,6 +39,7 @@ const initalPeerStatus = (t: TFunction) => ({
   dataLastSyncedOn: {
     text: '',
   },
+  replicationHealth: {},
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx:42]:
```suggestion
  replicationHealth: null,
```

File: packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx [hunk 17]
```diff
@@ -137,13 +150,28 @@ export const PeerClusterStatus: React.FC<PeerClusterStatusProps> = ({
...
+        peerCurrentStatus.peerReadiness.text ===
+        PEER_READINESS(t).PEER_NOT_READY
+      ) {
+        setErrorMessage(
+          actionType === DRActionType.FAILOVER
+            ? ErrorMessageType.FAILOVER_READINESS_CHECK_FAILED
+            : ErrorMessageType.RELOCATE_READINESS_CHECK_FAILED
+        );
+      } else if (
+        peerCurrentStatus.replicationHealth &&
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx:163]:
```suggestion
        !!peerCurrentStatus.replicationHealth &&
```

Comment (GowthamShanmugam) [reply packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx:163]: or you can ignore this condition by adding '?'  in  peerCurrentStatus.replicationHealth as VOLUME_REPLICATION_HEALTH

File: packages/mco/types/ramen.ts [hunk 18]
```diff
@@ -70,6 +70,7 @@ export type DRPlacementControlKind = K8sResourceCommon & {
     //  A list of namespaces that are protected by the DRPC.
     //  N/A for the managed  applications.
     protectedNamespaces?: string[];
+    schedulingInterval?: string;
```
Comment (GowthamShanmugam) [inline packages/mco/types/ramen.ts:73]:
```suggestion
  
```
there is no schedulingInterval in DRPlacementControlKind type

File: packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx [hunk 19]
```diff
@@ -78,25 +83,33 @@ const getPeerStatusSummary = (
...
-        : acc,
-    initalPeerStatus(t)
-  );
+  drpcStateList?.reduce((acc, drpcState) => {
+    const drPlacementControl = drpcState?.drPlacementControl;
+
+    if (subsGroups.includes(getName(drPlacementControl))) {
+      const lastGroupSyncTime = drPlacementControl?.status?.lastGroupSyncTime;
+      const volumesSchedulingInterval =
+        drPlacementControl?.spec?.schedulingInterval || '0m';
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx:92]:
```suggestion
     
```
you need to get the schedulingInterval from DRPolicy

**Comment by TimothyAsirJeyasing** [comment]
Screenshot from 2024-10-28 17-25-25

File: packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx [hunk 20]
```diff
@@ -75,28 +80,54 @@ const getPeerStatusSummary = (
...
-          }
-        : acc,
-    initalPeerStatus(t)
-  );
+  drpcStateList?.reduce((acc, drpcState) => {
+    const drPlacementControl = drpcState?.drPlacementControl;
+
+    if (subsGroups.includes(getName(drPlacementControl))) {
+      const lastGroupSyncTime = drPlacementControl?.status?.lastGroupSyncTime;
+      const volumesSchedulingInterval = schedulingInterval;
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx:91]:
```suggestion
```
Why this extra variable? we can use schedulingInterval directly

File: packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx [hunk 21]
```diff
@@ -75,28 +80,54 @@ const getPeerStatusSummary = (
...
+      return {
+        ...acc,
+        peerReadiness: getPeerReadiness(
+          acc.peerReadiness,
+          drpcState,
+          actionType,
+          t
+        ),
+        dataLastSyncedOn: getDataLastSyncTime(acc.dataLastSyncedOn, drpcState),
+        replicationHealth: higherSeverityHealth,
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx:114]:
```suggestion
        replicationHealth: getHigherSeverityHealth(
        acc.replicationHealth,
        lastGroupSyncTime,
        schedulingInterval,
      );
```

File: packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx [hunk 22]
```diff
@@ -75,28 +80,54 @@ const getPeerStatusSummary = (
...
+    if (subsGroups.includes(getName(drPlacementControl))) {
+      const lastGroupSyncTime = drPlacementControl?.status?.lastGroupSyncTime;
+      const volumesSchedulingInterval = schedulingInterval;
+      const replicationType = getReplicationType(volumesSchedulingInterval);
+
+      const currentHealth = getReplicationHealth(
+        lastGroupSyncTime,
+        volumesSchedulingInterval,
+        replicationType
+      );
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx:98]: lets move this inside getHigherSeverityHealth

File: packages/mco/components/modals/app-failover-relocate/subscriptions/error-messages.tsx [hunk 23]
```diff
@@ -17,6 +17,8 @@ export enum ErrorMessageType {
   PRIMARY_CLUSTER_IS_NOT_FENCED,
   TARGET_CLUSTER_IS_FENCED,
   NO_SUBSCRIPTION_GROUP_FOUND,
+  // Warning Messages
+  VOLUME_SYNC_DELAY = 20,
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/error-messages.tsx:21]: lets keep this 20

Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/error-messages.tsx:21]:
```suggestion
  VOLUME_SYNC_DELAY
```

File: packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx [hunk 24]
```diff
@@ -113,17 +144,25 @@ export const PeerClusterStatus: React.FC<PeerClusterStatusProps> = ({
...
   );
   const setErrorMessage = React.useCallback(
     (errorMessage: ErrorMessageType) => {
       dispatch({
         type: FailoverAndRelocateType.SET_ERROR_MESSAGE,
-        payload: {
-          peerStatusErrorMessage: errorMessage,
-        },
+        payload:
+          errorMessage < 20
```
Comment (GowthamShanmugam) [inline packages/mco/components/modals/app-failover-relocate/subscriptions/peer-cluster-status.tsx:161]:
```suggestion
          !errorMessage. VOLUME_SYNC_DELAY
```

**Review by GowthamShanmugam — APPROVED** [review]

**Pull request merged**